platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,290 @@
+// just as it does today.
+```
+
+### Bootstrapping Genesis for a brand new network
+
+This proposal already creates a new interface for the services to install a Candidate Roster—via a Platform API, or
+similar—which ultimately stores the Candidate Roster in a designated location in the State.
+The Startup procedure described above consumes the Candidate Roster from that location and starts its processing or
+adoption based on its readiness.
+
+The `config.txt` will only ever be used literally once—when a brand new network goes through its genesis.
+The `config.txt` will never be used again in the network life cycle.
+
+Given that:
+
+1. we need a special mechanism for supplying a genesis roster to a brand new network, and
+2. this mechanism is only ever used once during genesis, and
+3. we already have a support for `config.txt` in the platform, and
+4. we already propose a public interface for Services to install Candidate Rosters in existing networks
+
+there should be absolutely no need to introduce any additional mechanisms or interfaces between Platform and Services to
+supply a genesis roster.
+
+If we must, we may add a field to the `config.txt` to indicate the number of shares for each node.
+This should be a relatively trivial change.
+However, since the `config.txt` would only ever be used once during genesis, it’s a lot simpler to skip that and assume
+equal shares for all genesis nodes, and only start supporting non-equal shares when a new Candidate Roster is installed
+via the new public interface of the Platform API. But as mentioned, if we must support this at genesis for brand new
+networks, we can do that, and it should be trivial.
+
+### Roster changes needed for Components
+
+- Reconnect. Reconnect logic currently exchanges and validates address book between the learner and teacher nodes. The
+  learner node uses the address book to select the teacher node, as well as remove all invalid signatures from the
+  state. Both of these will need to be updated to use rosters instead.
+- Networking. Most parts of the network code have been abstracted away from the address book as part of the Dynamic
+  Address Book effort, so there should be minimal work left there.
+- Event Validation. The event validation logic uses the address book to determine whether a given event has a valid
+  signature. This will be updated to use rosters instead.
+- Miscellaneous components. Other components that reference things stored in the address book (like consensus needing
+  the node weight or the crypto module verifying keys) will need to be updated to use things stored in rosters instead.
+
+### Test Plan
+
+Some of the obvious test cases to be covered in the plan include validating one or more of the following scenarios:
+
+1. New valid Candidate Roster created with no subsequent one sent by app. Verify accept.
+2. New valid Candidate Roster created with subsequent one sent by app. Verify accept.
+3. Invalid roster(s) sent. Verify reject.
```

Comment:
What is the definition of `valid` roster?   That needs to be explained clearly and precisely in an earlier section.

Reply:
hmmm I'm not sure about including that actually. I think the presense of the protobuf/data structure implies what constitutes validity e.g. it must have >0 RosterEntry/ies, with each having non-zero weight with valid certificates etc. Do you think that's valuable information?

Reply:
Is something like this useful?


###### Roster Validity
In simple terms, the following constitutes a valid roster:
1. The roster must have at least one RosterEntry.
2. All RosterEntry/ies must have a non-zero weight.
3. All RosterEntry/ies must have a valid X509Certificate.
4. All RosterEntry/ies must have a valid PairingPublicKey.
5. All RosterEntry/ies must have at least one ServiceEndpoint.
6. All ServiceEndpoint/s must have a valid IP address, port, or domain name.
7. The roster must have a unique NodeId for each RosterEntry.